lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -4427,6 +4432,84 @@ INITIALIZE_PASS(SROA_Parameter_HLSL, "scalarrepl-param-hlsl",
   "Scalar Replacement of Aggregates HLSL (parameters)", false,
   false)
 
+void SROA_Parameter_HLSL::RewriteBitcastWithIdenticalStructs(Function *F) {
+  if (F->isDeclaration())
+    return;
+  // Gather list of bitcast involving src and dest structs with identical layout
+  std::vector<BitCastInst*> worklist;
+  for (inst_iterator I = inst_begin(F), E = inst_end(F); I != E; ++I) {
+    if (BitCastInst *BCI = dyn_cast<BitCastInst>(&*I)) {
+      Type *DstTy = BCI->getDestTy();
+      Type *SrcTy = BCI->getSrcTy();
+      if (!SrcTy->isPointerTy() || !DstTy->isPointerTy())
+        continue;
+      DstTy = DstTy->getPointerElementType();
+      SrcTy = SrcTy->getPointerElementType();
+      if (!SrcTy->isStructTy() || !DstTy->isStructTy())
+        continue;
+      StructType *DstST = cast<StructType>(DstTy);
+      StructType *SrcST = cast<StructType>(SrcTy);
+      if (!SrcST->isLayoutIdentical(DstST))
+        continue;
+      worklist.push_back(BCI);
+    }
+  }
+
+  // Replace bitcast involving src and dest structs with identical layout
+  while (!worklist.empty()) {
+    BitCastInst *BCI = worklist.back();
+    worklist.pop_back();
+    StructType *srcStTy = cast<StructType>(BCI->getSrcTy()->getPointerElementType());
+    StructType *destStTy = cast<StructType>(BCI->getDestTy()->getPointerElementType());
+    Value* srcPtr = BCI->getOperand(0);
+    IRBuilder<> AllocaBuilder(dxilutil::FindAllocaInsertionPt(BCI->getParent()->getParent()));
+    AllocaInst *destPtr = AllocaBuilder.CreateAlloca(destStTy);
+    IRBuilder<> InstBuilder(BCI);
+    std::vector<unsigned> idxlist = { 0 };
+    CopyElementsOfStructsWithIdenticalLayout(InstBuilder, destPtr, srcPtr, srcStTy, idxlist);
+    BCI->replaceAllUsesWith(destPtr);
+    BCI->eraseFromParent();
+  }
+}
+
+std::vector<Value *>
+SROA_Parameter_HLSL::GetConstValueIdxList(IRBuilder<> &builder,
+                                          std::vector<unsigned> idxlist) {
+  std::vector<Value *> idxConstList;
+  for (unsigned idx : idxlist) {
+    idxConstList.push_back(ConstantInt::get(builder.getInt32Ty(), idx));
+  }
+  return idxConstList;
+}
+
+void SROA_Parameter_HLSL::CopyElementsOfStructsWithIdenticalLayout(
+    IRBuilder<> &builder, Value *destPtr, Value *srcPtr, Type *ty,
+    std::vector<unsigned> idxlist) {
```

Comment:
Instead of taking the vector by value, can't you take it by ref, and pop_back after recursing? Your push/pops should be perfectly balanced.